src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -58,6 +59,93 @@ def check_type_refinement(node: foast.Expr, new: foast.DataType) -> None:
             )
 
 
+@dataclass
+class TypeInfo:
+
+    type: foast.SymbolType  # noqa: A003
+
+    @property
+    def is_complete(self) -> bool:
+        return is_complete_symbol_type(self.type)
+
+    @property
+    def is_any_type(self) -> bool:
+        if self.is_complete:
+            return False
+        return (self.type is None) or (self.type.constraint is None)
+
+    @property
+    def constraint(self) -> Optional[foast.SymbolType]:
+        if self.is_complete:
+            return self
+        elif not self.is_any:
+            return self.constraint
```

Comment:
Doesn't this create a infinite recursive call? Did you mean
```suggestion
            return self.type.constraint
```?

Reply:
It did, but I solved the problem by having `.constraint` not reference `.is_any_type`